inputremapper/injection/context.py

```diff
@@ -20,13 +20,54 @@
 
 
 """Stores injection-process wide information."""
+from typing import Callable, Awaitable, List, Dict, Optional
 
+import evdev
+from evdev.ecodes import EV_KEY
 
+from inputremapper.injection.consumers.mapping_handler import (
+    MappingHandler,
+    HierarchyHandler,
+    create_handler,
+     )
+from inputremapper.key import Key
 from inputremapper.logger import logger
 from inputremapper.injection.macros.parse import parse, is_this_a_macro
 from inputremapper.system_mapping import system_mapping
 from inputremapper.config import NONE, MOUSE, WHEEL, BUTTONS
 
+Callback = Callable[[evdev.InputEvent], Awaitable[bool]]
+
+
+def order_keys(keys: List[Key], key: Key) -> List[Key]:
+    """reorder the keys according to some improvised rules
+
+    such that a combination a+b+c is in front of a+b which is in front of b
+    for a+b+c vs. b+d+e  ¯\_(ツ)_/¯
+    in this example b would be the common key
+
+    Parameters
+    ----------
+    keys : List[Key]
+        the list which needs ordering
+    key: Key
+        the Key all members of Keys have in common
+    """
+    # TODO
```

Comment:
This would be the code responsible for making combinations behave correctly.
I am open for suggestions, especially what to do about `a+b+c` and `b+d+e` we should at least make sure it behaves always the same.

Reply:
what

Reply:
If the Key `b` ((1,48,1),) is pressed the event gets sent to all `MappingHandlers` which need to know about this key. 

> A mapping handler handles a single mapping in the preset e.g. `"1,30,1": ["a" , "keyboard"]` or `"1,30,1+1,31,1": ["BTN_LEFT" , "mouse"]`. It can be hierarchic (`KeysToKey`) or not (`HirarchyHandler`). For each event only one hierarchic handler can be notified, but all non hierarchic handlers will be notified.
> To achieve that we order all hierarchic handlers with `order_keys()` and wrap them with a `HirarchyHandler` 

By ordering the keys list we make sure the Handlers get grouped in a HirarchyHandler and called in the correct sequence such that the first handler which handles the event will suppress all following handlers.

Here is an Exampel:
 - We have 3 handlers for key "b":
   - `b` maps to "d"
   - `a+b` maps to "e"
   - `a+b+c` maps to "f"
 - the ordered list is
   - `[a+b+c, a+b, b]`
 - suppose we hold down key "a"
   - input-remapper injects whatever is mapped to "a"
 - and then we press key "b"
   - input-remapper notifyes all non hierarchic handlers which care about `b`
   - `HierarchyMapper` will walk trough the list and notify the handlers:
     - `a+b+c`: notify returns `False`  since c is not pressed
     - `a+b`: notify returns `True`
     - `b`: will be notified with `suppress = True` (let the handler know about the key-down but suppress its injection)

Reply:
That is the Idea

Reply:
Can you test out the combination handling? The new system currently only supports combination -> single key mappings. No macros or anything fancy yet.

There is now a different set of rules for combinations:
1. combinations trigger if at least all keys of the combination are pressed down
2. combinations release if **any** key of the combination is released
3. before triggering all forwarded keys which are part of the combination get released
4. we don't care about the order in which keys are pressed
**Overlapping Combinations:**
5. each key press can only ever trigger one combination
6. a key which completes multiple combinations will only trigger the longest combination
7. if there are multiple longest combinations:
     - take the sum of all keycodes for each combination
     - trigger the one which has the largest sum
     - this ensures we get always the same result otherwise we would be dependent on the order in which the mappings got saved to the preset.

especially with overlapping combinations I got a few times stuck keys on the forwarded device, this is dependent on the order in which keys get pressed and released. Hopefully it is fixed now.

I did not modify the preset file (yet) so the GUI still works

Reply:
mapping:

`1 + 2` -> `b`

1. 1 down -> 1 written
2. 2 down -> b written
3. up 1 and 2
4. 2 down -> b written
5. 2 up and down -> more bs written

When I trigger the combination again, it starts working again and "2" is written, like it should.

It happens randomly, maybe after half of all combinations

```
1633351 15:24:00.841503 SPAM keycode_mapper.py:459: ((1, 28, 0)) ------------------ forwarding release
1633351 15:24:01.593660 SPAM keycode_mapper.py:512: ((1, 2, 1)) ------------------- maps to macro (m(a,m(0,h())), keyboard)
1633351 15:24:01.604211 DEBUG keycode_mapper.py:321: macro_sending (1, 30, 1)
1633351 15:24:01.625029 DEBUG keycode_mapper.py:321: macro_sending (1, 11, 1)
1633351 15:24:01.713712 SPAM keycode_mapper.py:427: ((1, 2, 0)) ------------------- releasing macro
1633351 15:24:01.713826 SPAM keycode_mapper.py:440: ((1, 2, 0)) ------------------- releasing key
1633351 15:24:01.724231 DEBUG keycode_mapper.py:321: macro_sending (1, 11, 0)
1633351 15:24:01.745016 DEBUG keycode_mapper.py:321: macro_sending (1, 30, 0)
1633351 15:24:02.177615 SPAM keycode_mapper.py:539: ((1, 57, 1)) ------------------ forwarding
1633351 15:24:02.297753 SPAM keycode_mapper.py:459: ((1, 57, 0)) ------------------ forwarding release
1633351 15:24:02.457766 SPAM keycode_mapper.py:512: ((1, 2, 1)) ------------------- maps to macro (m(a,m(0,h())), keyboard)
1633351 15:24:02.468133 DEBUG keycode_mapper.py:321: macro_sending (1, 30, 1)
1633351 15:24:02.488848 DEBUG keycode_mapper.py:321: macro_sending (1, 11, 1)
1633351 15:24:02.657657 SPAM mapping_handler.py:133: Key((1, 2, 1), (1, 3, 1)) ----- maps to ([(1, 48, 1), 'keyboard'])
1633351 15:24:02.657786 DEBUG mapping_handler.py:87: injecting form new mapping handler: (1, 48, 1)
1633351 15:24:02.865689 SPAM mapping_handler.py:133: Key((1, 2, 1), (1, 3, 1)) ----- maps to ([(1, 48, 0), 'keyboard'])
1633351 15:24:02.865804 DEBUG mapping_handler.py:87: injecting form new mapping handler: (1, 48, 0)
1633351 15:24:02.866097 SPAM keycode_mapper.py:469: ((1, 3, 0)) ------------------- unexpected key up
1633351 15:24:03.121736 SPAM keycode_mapper.py:389: ((1, 2, 1), (1, 57, 1)) ------- unknown combination
1633351 15:24:03.121839 SPAM keycode_mapper.py:539: ((1, 57, 1)) ------------------ forwarding
1633351 15:24:03.217768 SPAM keycode_mapper.py:459: ((1, 57, 0)) ------------------ forwarding release
1633351 15:24:03.305816 SPAM keycode_mapper.py:531: ((1, 2, 1), (1, 3, 1)) -------- maps to (48, keyboard)
1633351 15:24:03.401880 SPAM keycode_mapper.py:446: ((1, 3, 0)) ------------------- releasing (48, keyboard)
1633351 15:24:03.521923 SPAM keycode_mapper.py:531: ((1, 2, 1), (1, 3, 1)) -------- maps to (48, keyboard)
1633351 15:24:03.617858 SPAM keycode_mapper.py:446: ((1, 3, 0)) ------------------- releasing (48, keyboard)
1633351 15:24:03.705938 SPAM keycode_mapper.py:531: ((1, 2, 1), (1, 3, 1)) -------- maps to (48, keyboard)
1633351 15:24:03.849909 SPAM keycode_mapper.py:446: ((1, 3, 0)) ------------------- releasing (48, keyboard)
```

Reply:
At the bottom of the log it should say

```
1665743 15:33:37.631813 SPAM keycode_mapper.py:539: ((1, 3, 1)) ------------------- forwarding
1665743 15:33:37.751806 SPAM keycode_mapper.py:459: ((1, 3, 0)) ------------------- forwarding release
```

instead

Reply:
oooh that is unfortunate. The old keycode-mapper is doing stuff which is messing everything up. I should have expected that. Will remove it from the consumer-control

Reply:
now it should work.

Reply:
`1 + 2` -> `a`
`2 + 1` -> `b`

```
2060570 17:00:20.303041 SPAM mapping.py:217: Key((1, 2, 1), (1, 3, 1)) maps to ('a', 'keyboard')
2060570 17:00:20.303098 SPAM mapping.py:217: Key((1, 3, 1), (1, 2, 1)) maps to ('b', 'keyboard')
```

pressing 1 and 2 prints:

```
2060709 17:01:39.746127 SPAM mapping_handler.py:133: Key((1, 2, 1), (1, 3, 1)) ----- maps to ([(1, 30, 1), 'keyboard'])
2060709 17:01:39.746269 DEBUG mapping_handler.py:87: injecting form new mapping handler: (1, 30, 1)
2060709 17:01:39.922219 SPAM mapping_handler.py:133: Key((1, 3, 1), (1, 2, 1)) ----- maps to ([(1, 48, 0), 'keyboard'])
2060709 17:01:39.922365 DEBUG mapping_handler.py:87: injecting form new mapping handler: (1, 48, 0)
2060709 17:01:39.922639 SPAM mapping_handler.py:133: Key((1, 2, 1), (1, 3, 1)) ----- maps to ([(1, 30, 0), 'keyboard'])
2060709 17:01:39.922712 DEBUG mapping_handler.py:87: injecting form new mapping handler: (1, 30, 0)

Reply:
So far this is by design. But a good point, I should change it. And I need to find a way to supress all injection by the mapping which did not trigger.

Reply:
okay, now `1+2` will actually write a, and `2+1` will write b
Also we no longer send unnecessary release events to the mapped preset.

there are still undefined edge cases like:
`1+2` -> `a`
`3+2` -> `b`
what will happen when pressing `1+3+2`? Either `a` or `b` will be sent, this is currently undefined behavior. Most likely `b` should be triggered. Resolving this would require us to keep track of the order of all currently pressed keys, and even then I am not sure if it is possible with my current approach.

Reply:
I also think that `b` would be correct. But no strong feelings on my side.

What should work is that if only `1 + 2 + 3` -> `a` is mapped, and `0 + 2 + 1 + 3` is typed, the result should be "021a". Everything else is probably also undefined in the current behavior

Reply:
> What should work is that if only 1 + 2 + 3 -> a is mapped, and 0 + 2 + 1 + 3 is typed, the result should be "021a". Everything else is probably also undefined in the current behavior

that is the case with the new injection. 

I am quiet happy with the combination handling as it is now. I would leave it like that if there are no objections.